lib/fetch/index.js

```diff
@@ -740,34 +742,86 @@ function finalizeResponse (fetchParams, response) {
+  }
+
+  // 2. Let processResponseEndOfBody be the following steps:
+  const processResponseEndOfBody = () => {
+    // 1. Set fetchParams’s request’s done flag.
+    fetchParams.request.done = true
 
-  // 1. If fetchParams’s process response is non-null,
-  // then queue a fetch task to run fetchParams’s process response
-  // given response, with fetchParams’s task destination.
+    // 2. If fetchParams’s process response end-of-body is not null, then queue a fetch task to run fetchParams’s
+    //    process response end-of-body given response with fetchParams’s task destination.
+    if (fetchParams.processResponseEndOfBody != null) {
+      // TODO (spec): what exactly are we supposed to do with fetchParams.taskDestination?
+      fetchParams.processResponseEndOfBody(response)
+    }
+  }
+
+  // 3. If fetchParams’s process response is non-null, then queue a fetch task to run fetchParams’s process response given response,
+  //    with fetchParams’s task destination.
   if (fetchParams.processResponse != null) {
     fetchParams.processResponse(response)
   }
 
-  // 2. If fetchParams’s process response consume is non-null, then:.
-  //    TODO
-  //    1. Let processBody given nullOrBytes be this step: run fetchParams’s
-  //    process response consume given response and nullOrBytes.on.
-  //    TODO
-  //    2. Let processBodyError be this step: run fetchParams’s process
-  //    response consume given response and failure.on.
-  //    TODO
-  //    3. If response’s body is null, then queue a fetch task to run
-  //    processBody given null, with fetchParams’s task destination.on.
-  //    TODO
-  //    4. Otherwise, fully read response’s body given processBody,
-  //    processBodyError, and fetchParams’s task destination.on.
-  //    TODO
-
-  // TODO (spec): The spec doesn't specify this but we need to
-  // terminate fetch if we have an error response.
-  if (response.type === 'error') {
-    context.terminate({ reason: response.error })
+  // 4. If response’s body is null, then run processResponseEndOfBody.
+  if (response.body == null) {
+    processResponseEndOfBody()
+  }
+
+  // 5. Otherwise:
+  else {
+    TransformStream ??= require('stream/web').TransformStream
+    // 1. Let transformStream be a new a TransformStream.
+
+    // 2. Let identityTransformAlgorithm be an algorithm which, given chunk, enqueues chunk in transformStream.
+    const identityTransformAlgorithm = (chunk, controller) => {
+      controller.enqueue(chunk)
+    }
+
+    const transformStream = new TransformStream({
+      // 3. Set up transformStream with transformAlgorithm set to identityTransformAlgorithm and
+      //    flushAlgorithm set to processResponseEndOfBody.
+      transform: identityTransformAlgorithm,
+      flush: processResponseEndOfBody
+    })
+
+    // 4. Set response’s body to the result of piping response’s body through transformStream.
+    // Khafra note: although the spec is unclear, only readable streams can be piped through
```

Comment:
Is there a case where the stream is not readable here? @dominic

Reply:
I don't believe so, but when I was working on this pr I was getting a lot of test failures so I added the readable check just in case. Removing it, the tests seem to pass so I have gone ahead and removed it in def5a6be2e2b6a6d7767c260d7ceab666ccc3201